connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/HerderRequestHandler.java

```diff
@@ -53,9 +53,30 @@ public void requestTimeoutMs(long requestTimeoutMs) {
     }
 
     /**
-     * Wait for a FutureCallback to complete. If it succeeds, return the parsed response. If it fails, try to forward the
-     * request to the leader.
-      */
+     * Wait for a {@link FutureCallback} to complete and return the result if successful.
+     * @param cb the future callback to wait for
+     * @return the future callback's result if successful
+     * @param <T> the future's result type
+     * @throws Throwable if the future callback isn't successful
+     */
+    public <T> T completeRequest(FutureCallback<T> cb) throws Throwable {
+        try {
+            return cb.get(requestTimeoutMs, TimeUnit.MILLISECONDS);
+        } catch (ExecutionException e) {
+            throw e.getCause();
+        } catch (TimeoutException e) {
+            // This timeout is for the operation itself. None of the timeout error codes are relevant, so internal server
+            // error is the best option
+            throw new ConnectRestException(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(), "Request timed out");
+        } catch (InterruptedException e) {
+            throw new ConnectRestException(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(), "Request interrupted");
+        }
+    }
```

Comment:
I'm not sure we really need this variant. Sure, there are some asynchronous requests that don't require forwarding to the leader or a specific worker at the moment, but if we have to change the logic for them to start doing forwarding, it's easier to just tweak the `DistributedHerder` class without having to remember to also tweak `ConnectorsResource`.

Reply:
Fair enough! I wasn't really thinking about the URL/HTTP verb/etc. parameters but that's a good point.

In that case, two suggestions:

1. Can we leverage this variant in all the other appropriate places in `ConnectorsResource` where forwarding isn't necessary but an asynchronous `Herder` API is used?
2. Would it be possible to leverage `completeRequest` inside `completeOrForwardRequest` so that we reduce the duplicated logic for handling timeouts, interruptions, specifying a request timeout, etc.?